llama_index/node_parser/postprocessors/metadata_extractors.py

```diff
@@ -0,0 +1,393 @@
+        self._extractors: List[Callable[[Sequence[BaseNode]], None]] = []
+        self._node_text_template = node_text_template
+        self._disable_template_rewrite = disable_template_rewrite
+
+    def extract_title(self, **kwargs: Any) -> "MetadataExtractor":
+        """Extract document title for the sequence of nodes.
+        Args:
+            nodes (int): number of nodes (from front of sequence)
+                to use for title extraction
+            node_template (str): template for node-level title
+                clues extraction
+            combine_template (str): template for combining node-level
+                clues into a document-level title
+        """
+        self._extractors.append(
+            TitleExtractor(llm_predictor=self._llm_predictor, **kwargs)
+        )
+        return self
+
+    def extract_questions_answered(self, **kwargs: Any) -> "MetadataExtractor":
+        """Extract questions answered by the node.
+        Args:
+            questions (int): number of questions to extract
+            prompt_template (str): template for question extraction
+        """
+        self._extractors.append(
+            QuestionsAnsweredExtractor(llm_predictor=self._llm_predictor, **kwargs)
+        )
+        return self
+
+    def extract_keywords(self, **kwargs: Any) -> "MetadataExtractor":
+        """Extract keywords for the node.
+        Args:
+            keywords (int): number of keywords to extract
+        """
+        self._extractors.append(
+            KeywordExtractor(llm_predictor=self._llm_predictor, **kwargs)
+        )
+        return self
+
+    def extract_summary(self, **kwargs: Any) -> "MetadataExtractor":
+        """Extract summary for the node.
+        Args:
+            summaries (List[str]): list of summaries to extract: "self", "prev", "next"
+            prompt_template (str): template for summary extraction"""
+        self._extractors.append(
+            SummaryExtractor(llm_predictor=self._llm_predictor, **kwargs)
+        )
+        return self
+
+    def extract(
+        self, extractor: Callable[[Sequence[BaseNode]], None]
+    ) -> "MetadataExtractor":
+        """Extract metadata using a custom extractor.
+        Args:
+            extractor (Callable[[Sequence[BaseNode]], None]):
+                Callable that takes a sequence of nodes and mutates the
+                metadata of the nodes in place. It can be a custom class with an
+                  appropriate __call__ or a simply a function/lambda function.
+        """
+        self._extractors.append(extractor)
+        return self
+
+    def build(self) -> "MetadataAugmentationPostProcessor":
+        return MetadataAugmentationPostProcessor(
+            self._extractors, self._node_text_template, self._disable_template_rewrite
+        )
+
+
+class MetadataAugmentationPostProcessor(NodeParserPostProcessor):
+    def __init__(
+        self,
+        extractors: Sequence[Callable[[Sequence[BaseNode]], None]],
```

Comment:
seems like typing should be List[MetadataExtractorBase]

Reply:
I wasn't sure if we should require the metadata extractor to be a subclass of `MetadataExtractorBase` or just an arbitrary callable. I find defining a function to be less cumbersome, but I'm open to just forcing the user to implement `MetadataExtractorBase`.

Reply:
yeah we should be restrictive for clarity unless we have a reason otherwise